net: ipv6: Improve routing between interfaces

kderda opened the pull request:
This PR includes few fixes for routing IPv6 packet between interfaces. Every change is described in in commit's message.

kderda committed:
net: ipv6: routing: match destination address to a given interface

Before packet is routed there is a check verifying if the destination
address is not the device's one. However, the check should be limited
to the packet's original interface. Otherwise, packet cannot be routed
if it is destined for an other interface of the device.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

kderda committed:
net: ipv6: routing: do not use an original source address for NS

When a packet's transmission is prepared and Neighbor Solicitation is
sent its souce address is always the same as of a packet awaiting -
also for packets routed from an other interface.

Quote from the RFC:
If the source address of the packet prompting the solicitation is the
same as one of the addresses assigned to the outgoing interface, that
address SHOULD be placed in the IP Source Address of the outgoing
solicitation.  Otherwise, any one of the addresses assigned to the
interface should be used.

This commit fixes the behavior.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

kderda committed:
net: ipv6: routing: do not loop back already routed packets

If the packet was routed between interfaces by IPv6 module it should
not be looped back but has to be passed to the destination interface
instead.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

kderda committed:
net: ipv6: routing: move checking for an own source address

When IPv6 packet is received, there is a check of the packet's source
address to verify that it is not interface's non-tentative address.

This commit moves this check to the later stages of processing as
packets that can be routed are dropped in the early stage otherwise.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

jukkar reviewed (dismissed):
LGTM

pdgendt reviewed (dismissed):
Nit: fix the compliance issues

rlubos committed:
tests: net: sockets: misc: Fix SO_BINDTODEVICE test

The test case for SO_BINDTODEVICE socket option was flaky, the client
socket always sent the datagram to the IP address of the second
interface, so in theory every packet should end up on that interface.
In practice though, due to imperfect loopback packet handling, the test
worked as the packet ended up on the interface it was sent from.

The test should send datagrams to the IP addresses of the interface 1
and 2 alternatively. The server socket binds to ANY address, so w/o
interface binding it should receive all datagrams, so it allows to
verify if SO_BINDTODEVICE filtering works fine.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

konradderda commented:
@rlubos wow, thanks a lot! I have cherry-picked the commit.

rlubos commented on tests/net/socket/misc/src/main.c:
```diff
@@ -310,23 +309,23 @@ void test_ipv4_so_bindtodevice(void)
-	zassert_equal(ret, 1, "inet_pton failed");
+	peer_addr_1.sin_family = AF_INET;
+	peer_addr_1.sin_port = htons(DST_PORT);
+	peer_addr_1.sin_addr = my_ipv4_addr1;
```
There's always something... It seems that userspace doesn't allow to reuse global variables for initialization... Weird it didn't manifest itself on native_sim, only on qemu platforms. TBH I'm not sure if there's a point digging into this, I'd fix it like this:
```patch
diff --git a/tests/net/socket/misc/src/main.c b/tests/net/socket/misc/src/main.c
index 1e22dada30e..a245c0f64e4 100644
--- a/tests/net/socket/misc/src/main.c
+++ b/tests/net/socket/misc/src/main.c
@@ -297,7 +297,16 @@ void test_ipv4_so_bindtodevice(void)
 {
        int sock_c;
        int sock_s;
-       struct sockaddr_in peer_addr_1, peer_addr_2;
+       struct sockaddr_in peer_addr_1 = {
+               .sin_family = AF_INET,
+               .sin_port = htons(DST_PORT),
+               .sin_addr = { { { 192, 0, 2, 1 } } },
+       };
+       struct sockaddr_in peer_addr_2 = {
+               .sin_family = AF_INET,
+               .sin_port = htons(DST_PORT),
+               .sin_addr = { { { 192, 0, 2, 2 } } },
+       };
        struct sockaddr_in bind_addr = {
                .sin_family = AF_INET,
                .sin_port = htons(DST_PORT),
@@ -309,13 +318,6 @@ void test_ipv4_so_bindtodevice(void)
        sock_s = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
        zassert_true(sock_s >= 0, "socket open failed");
 
-       peer_addr_1.sin_family = AF_INET;
-       peer_addr_1.sin_port = htons(DST_PORT);
-       peer_addr_1.sin_addr = my_ipv4_addr1;
-       peer_addr_2.sin_family = AF_INET;
-       peer_addr_2.sin_port = htons(DST_PORT);
-       peer_addr_2.sin_addr = my_ipv4_addr2;
-
        test_so_bindtodevice(sock_c, sock_s, (struct sockaddr *)&peer_addr_1,
                             (struct sockaddr *)&peer_addr_2, sizeof(peer_addr_1),
                             (struct sockaddr *)&bind_addr, sizeof(bind_addr));
@@ -325,7 +327,18 @@ void test_ipv6_so_bindtodevice(void)
 {
        int sock_c;
        int sock_s;
-       struct sockaddr_in6 peer_addr_1, peer_addr_2;
+       struct sockaddr_in6 peer_addr_1 = {
+               .sin6_family = AF_INET6,
+               .sin6_port = htons(DST_PORT),
+               .sin6_addr = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+                                  0, 0, 0, 0, 0, 0, 0, 0x1 } } },
+       };
+       struct sockaddr_in6 peer_addr_2 = {
+               .sin6_family = AF_INET6,
+               .sin6_port = htons(DST_PORT),
+               .sin6_addr = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+                                  0, 0, 0, 0, 0, 0, 0, 0x2 } } },
+       };
        struct sockaddr_in6 bind_addr = {
                .sin6_family = AF_INET6,
                .sin6_port = htons(DST_PORT),
@@ -337,13 +350,6 @@ void test_ipv6_so_bindtodevice(void)
        sock_s = zsock_socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
        zassert_true(sock_s >= 0, "socket open failed");
 
-       peer_addr_1.sin6_family = AF_INET6;
-       peer_addr_1.sin6_port = htons(DST_PORT);
-       peer_addr_1.sin6_addr = my_ipv6_addr1;
-       peer_addr_2.sin6_family = AF_INET6;
-       peer_addr_2.sin6_port = htons(DST_PORT);
-       peer_addr_2.sin6_addr = my_ipv6_addr2;
-
        test_so_bindtodevice(sock_c, sock_s, (struct sockaddr *)&peer_addr_1,
                             (struct sockaddr *)&peer_addr_2, sizeof(peer_addr_1),
                             (struct sockaddr *)&bind_addr, sizeof(bind_addr));
```
